tools/pvacvector/run.py

```diff
@@ -176,36 +176,35 @@ def create_graph(iedb_results, seq_tuples, spacers):
             elif not Paths.has_edge(ID_1, ID_2):
                 Paths.add_edge(ID_1, ID_2, weight=worst_case, spacer=spacer)
 
-    print("Graph contains " + str(len(Paths)) + " nodes and " + str(Paths.size()) + " edges.")
+    print("Graph contains " + str(len(Paths)) + " nodes (peptides) and " + str(Paths.size()) + " edges (junctions).")
     return Paths
 
 def check_graph_valid(Paths, seq_dict):
-    n_nodes_without_outgoing_edges = 0
-
     if len(Paths.nodes()) < len(seq_dict.keys()):
-        return False
+        return (False, "No valid junctions found for peptides: {}".format(set(seq_dict.keys()) - set(Paths.nodes())))
 
+    nodes_without_outgoing_edges = []
     for node in Paths.nodes():
         if len(Paths.out_edges(node)) == 0:
-            n_nodes_without_outgoing_edges += 1
-    if n_nodes_without_outgoing_edges > 1:
-        return False
+            nodes_without_outgoing_edges.append(node)
+    if len(nodes_without_outgoing_edges) > 1:
+        return (False, "More than one peptide without valid outgoing junction: {}".format(nodes_without_outgoing_edges))
 
-    n_nodes_without_incoming_edges = 0
+    nodes_without_incoming_edges = []
     for node in Paths.nodes():
         if len(Paths.in_edges(node)) == 0:
-            n_nodes_without_incoming_edges += 1
-    if n_nodes_without_incoming_edges > 1:
-        return False
+            nodes_without_incoming_edges.append(node)
+    if len(nodes_without_incoming_edges) > 1:
+        return (False, "More than one peptide without valid incoming junction: {}".format(nodes_without_incoming_edges))
 
-    n_nodes_without_any_edges = 0
+    nodes_without_any_edges = []
     for node in Paths.nodes():
         if len(Paths.in_edges(node)) == 0 and len(Paths.out_edges(node)) == 0:
-            n_nodes_without_any_edges += 1
-    if n_nodes_without_any_edges > 0:
-        return False
+            nodes_without_any_edges.append(node)
+    if len(nodes_without_any_edges) > 0:
+        return (False, "No valid junctions found for peptides: {}".format(nodes_without_any_edges))
```

Comment:
Is it possible to reach this `return`? Seems like the condition here would've necessarily triggered both of the previous errors.

Would it be beneficial to the user to go through the nodes populating both  `nodes_without_outgoing_edges` and `nodes_without_incoming_edges`, then checking and reporting on both together?

Reply:
In the two previous checks you could've ended up with one node without incoming edges and one node without outgoing edges. If those two nodes are the same node, this condition would trigger. However, I don't _think_ we would even add this node to the graph to begin with so this would get caught by the very first check. I'll probably leave it in just to be sure.

Yeah, it would be best to report on the full set.

Reply:
Ah! What I was missing here was the `0` in this condition versus the `1` in the others... so single node is still caught here.  Makes sense, thanks! :smile: